File: frontend/src/api/exchange.ts
import request from './axios'

// ==================== 类型定义 ====================

// 通用成功响应。
export interface SuccessResponse {
  code?: number
  message: string
}

// 商品数据结构（与后端 `models.Product` 的 JSON 字段对齐）。
export interface Product {
  id: number
  prize_id: string
  prize_name: string
  p_order: number
  category: string
  daily_remainder_count: number
  daily_limit_count?: number
  stock_status?: string
  last_stock_check?: string
  memo?: string
  is_active?: boolean
  is_deleted?: boolean
  created_at: string
  updated_at: string

  // 兼容历史字段（前端旧代码可能仍在引用）。
  prized_name?: string
  porder?: number
  prize_image?: string
  prize_price?: number
  prize_count?: number
}

// 兑换账号数据结构。
export interface ExchangeAccount {
  id: number
  user_id: number
  account_id: number
  phone: string
  remark: string
  exchange_time_1: string
  exchange_time_2: string
  is_active: boolean
  auth?: string
  token?: string
  jwt_token?: string
  last_exchange_at?: string
  created_at: string
  updated_at: string

  // 扩展关联字段（按后端预加载情况可能存在）。
  product_id?: number
  product?: Product
  account?: Record<string, unknown>
}

// 抢兑任务状态。
export type ExchangeTaskStatus = 'pending' | 'running' | 'completed' | 'cancelled' | 'failed'

// 抢兑任务类型。
export type ExchangeTaskType = 'fixed' | 'long_term'

// 抢兑任务数据结构。
export interface ExchangeTask {
  id: number
  user_id: number
  exchange_account_id: number
  product_id: number
  prize_id: string
  prize_name: string
  task_type: ExchangeTaskType
  max_attempts: number
  attempted_count: number
  status: ExchangeTaskStatus
  last_attempt_at?: string
  last_result?: string

  // 增强字段。
  priority?: number
  task_group?: string
  timeout_seconds?: number
  max_retries?: number
  retry_count?: number
  last_retry_at?: string
  success_count?: number
  fail_count?: number

  created_at: string
  updated_at: string
  exchange_account?: ExchangeAccount
  product?: Product
}

// 抢兑记录数据结构。
export interface ExchangeRecord {
  id: number
  user_id: number
  exchange_account_id: number
  exchange_task_id?: number
  product_id: number
  prize_id: string
  prize_name: string
  status: 'success' | 'failed'
  message: string
  execution_time_ms: number
  created_at: string

  // 预加载关联字段（记录列表页会使用）。
  product?: Product
  exchange_account?: ExchangeAccount
}

// 抢兑记录统计数据。
export interface RecordStats {
  success: number
  failed: number
}

// 商品搜索响应。
export interface SearchProductsResponse {
  products: Product[]
  total: number
}

// 商品分类响应。
export interface GetProductCategoriesResponse {
  categories: string[]
}

// 手动更新商品响应。
export interface UpdateProductsResponse extends SuccessResponse {
  data?: {
    account_id: number
    count: number
  }
}

// 获取兑换账号列表响应。
export interface GetExchangeAccountsResponse {
  accounts: ExchangeAccount[]
  total: number
}

// 添加兑换账号响应。
export interface AddExchangeAccountResponse {
  account: ExchangeAccount
}

// 获取抢兑任务列表响应。
export interface GetExchangeTasksResponse {
  tasks: ExchangeTask[]
  total: number
}

// 创建抢兑任务响应。
export interface CreateExchangeTaskResponse {
  task: ExchangeTask
}

// 抢兑记录响应。
export interface GetExchangeRecordsResponse {
  records: ExchangeRecord[]
  total: number
  stats: RecordStats
}

// 批量执行单个任务结果。
export interface BatchExecuteResult {
  task_id: number
  success: boolean
  message: string
}

// 批量执行抢兑任务响应。
export interface BatchExecuteExchangeTasksResponse {
  message: string
  results: BatchExecuteResult[]
}

// 抢兑配置数据结构（管理员）。
export interface ExchangeConfig {
  auto_update_products: boolean
  concurrency: number
  enabled: boolean
  exchange_monthly_enabled: boolean
  exchange_time: string
  monthly_prize_id: string
}

// ==================== 请求参数 ====================

// 添加兑换账号请求参数。
export interface AddExchangeAccountRequest {
  account_id: number
  remark?: string
  exchange_time_1?: string
  exchange_time_2?: string

  // 当前后端暂未消费该字段，但前端表单仍保留。
  product_id?: number
}

// 更新兑换账号请求参数。
export interface UpdateExchangeAccountRequest {
  remark?: string
  exchange_time_1?: string
  exchange_time_2?: string
  is_active?: boolean
}

// 创建抢兑任务请求参数。
export interface CreateExchangeTaskRequest {
  exchange_account_id: number
  product_id: number
  task_type?: ExchangeTaskType
  max_attempts?: number
}

// 更新抢兑任务请求参数。
export interface UpdateExchangeTaskRequest {
  max_attempts?: number
}

// 更新抢兑配置请求参数。
export interface UpdateExchangeConfigRequest {
  auto_update_products: boolean
  concurrency: number
  enabled: boolean
  exchange_monthly_enabled?: boolean
  exchange_time?: string
  monthly_prize_id?: string
}

// 抢兑记录查询参数。
export interface GetExchangeRecordsParams {
  page?: number
  limit?: number
  account_id?: number
  product_name?: string
  status?: 'success' | 'failed'
  start_date?: string
  end_date?: string
}

// 导出抢兑记录参数。
export interface ExportExchangeRecordsParams {
  account_id?: number
  product_name?: string
  status?: string
  start_date?: string
  end_date?: string
  format?: 'csv' | 'json'
}

// ==================== API 函数 ====================

// 搜索商品。
export function searchProducts(keyword: string, limit?: number): Promise<SearchProductsResponse> {
  return request<SearchProductsResponse>({
    url: '/api/products/search',
    method: 'get',
    params: { keyword, limit }
  })
}

// 获取商品分类。
export function getProductCategories(): Promise<GetProductCategoriesResponse> {
  return request<GetProductCategoriesResponse>({
    url: '/api/products/categories',
    method: 'get'
  })
}

// 手动更新商品。
export function updateProducts(accountId?: number): Promise<UpdateProductsResponse> {
  return request<UpdateProductsResponse>({
    url: '/api/products/update',
    method: 'post',
    data: { account_id: accountId }
  })
}

// 获取兑换账号列表。
export function getExchangeAccounts(): Promise<GetExchangeAccountsResponse> {
  return request<GetExchangeAccountsResponse>({
    url: '/api/exchange/accounts',
    method: 'get'
  })
}

// 添加兑换账号。
export function addExchangeAccount(data: AddExchangeAccountRequest): Promise<AddExchangeAccountResponse> {
  return request<AddExchangeAccountResponse>({
    url: '/api/exchange/accounts',
    method: 'post',
    data
  })
}

// 更新兑换账号。
export function updateExchangeAccount(id: number, data: UpdateExchangeAccountRequest): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: `/api/exchange/accounts/${id}`,
    method: 'put',
    data
  })
}

// 删除兑换账号。
export function deleteExchangeAccount(id: number): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: `/api/exchange/accounts/${id}`,
    method: 'delete'
  })
}

// 创建抢兑任务。
export function createExchangeTask(data: CreateExchangeTaskRequest): Promise<CreateExchangeTaskResponse> {
  return request<CreateExchangeTaskResponse>({
    url: '/api/exchange/tasks',
    method: 'post',
    data
  })
}

// 获取抢兑任务列表。
export function getExchangeTasks(): Promise<GetExchangeTasksResponse> {
  return request<GetExchangeTasksResponse>({
    url: '/api/exchange/tasks',
    method: 'get'
  })
}

// 更新抢兑任务。
export function updateExchangeTask(id: number, data: UpdateExchangeTaskRequest): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: `/api/exchange/tasks/${id}`,
    method: 'put',
    data
  })
}

// 删除抢兑任务。
export function deleteExchangeTask(id: number): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: `/api/exchange/tasks/${id}`,
    method: 'delete'
  })
}

// 立即执行抢兑任务。
export function executeExchangeTask(id: number): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: `/api/exchange/tasks/${id}/execute`,
    method: 'post'
  })
}

// 批量执行抢兑任务。
export function batchExecuteExchangeTasks(taskIds: number[]): Promise<BatchExecuteExchangeTasksResponse> {
  return request<BatchExecuteExchangeTasksResponse>({
    url: '/api/exchange/tasks/batch-execute',
    method: 'post',
    data: { task_ids: taskIds }
  })
}

// 获取抢兑配置（管理员）。
export function getExchangeConfig(): Promise<ExchangeConfig> {
  return request<ExchangeConfig>({
    url: '/api/admin/exchange/config',
    method: 'get'
  })
}

// 更新抢兑配置（管理员）。
export function updateExchangeConfig(data: UpdateExchangeConfigRequest): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: '/api/admin/exchange/config',
    method: 'put',
    data
  })
}

// 立即执行月卡兑换（管理员）。
export function executeMonthlyExchange(): Promise<SuccessResponse> {
  return request<SuccessResponse>({
    url: '/api/admin/exchange/execute-monthly',
    method: 'post'
  })
}

// 查询抢兑记录。
export function getExchangeRecords(params: GetExchangeRecordsParams): Promise<GetExchangeRecordsResponse> {
  return request<GetExchangeRecordsResponse>({
    url: '/api/exchange/records',
    method: 'get',
    params
  })
}

// 导出抢兑记录。
export function exportExchangeRecords(params: ExportExchangeRecordsParams): Promise<Blob> {
  return request<Blob>({
    url: '/api/exchange/records/export',
    method: 'get',
    params,
    responseType: 'blob'
  })
}

File: backend/internal/models/exchange.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Product 商品模型
type Product struct {
	ID                  uint           `gorm:"primarykey" json:"id"`
	PrizeID             string         `gorm:"uniqueIndex;size:100;not null" json:"prize_id"`         // 商品 ID(来自 API)
	PrizeName           string         `gorm:"column:prize_name;size:255;not null" json:"prize_name"` // 商品名称
	POrder              int            `gorm:"default:0" json:"p_order"`                              // 云朵价格
	Category            string         `gorm:"size:100;default:'未知分类'" json:"category"`               // 商品分类
	DailyRemainderCount int            `gorm:"default:0" json:"daily_remainder_count"`                // 每日剩余数量
	DailyLimitCount     int            `gorm:"default:0" json:"daily_limit_count"`                    // 每日限购数量
	DailyCount          int            `gorm:"default:0" json:"daily_count"`                          // 每日发放上限
	ImageURL            string         `gorm:"column:image_url;size:500;default:''" json:"image_url"` // 商品图片URL
	StockStatus         string         `gorm:"size:20;default:'unknown'" json:"stock_status"`         // 库存状态
	LastStockCheck      *time.Time     `json:"last_stock_check,omitempty"`                            // 最后库存检查时间
	Memo                string         `gorm:"type:text" json:"memo"`                                 // 商品备注信息
	IsActive            bool           `gorm:"default:true" json:"is_active"`                         // 是否启用
	IsDeleted           bool           `gorm:"default:false" json:"is_deleted"`                       // 是否已删除（软删除标记）
	UpdatedAt           time.Time      `json:"updated_at"`
	CreatedAt           time.Time      `json:"created_at"`
	DeletedAt           gorm.DeletedAt `gorm:"index" json:"-"`
}

// ExchangeRule 抢兑规则模型。
//
// 历史 API/前端仍会看到 exchange_account_id / exchange_account 这些兼容字段，
// 但物理表已经统一为 exchange_rules，任务和记录通过 exchange_rule_id 关联。
type ExchangeRule struct {
	ID             uint           `gorm:"primarykey" json:"id"`
	UserID         uint           `gorm:"not null;index" json:"user_id"`                                              // 用户 ID
	AccountID      uint           `gorm:"not null;index" json:"account_id"`                                           // 云盘账号 ID
	Phone          string         `gorm:"size:20;not null" json:"phone"`                                              // 手机号
	Auth           string         `gorm:"type:text;not null" json:"-"`                                                // Basic Auth 字符串
	Token          string         `gorm:"type:text" json:"-"`                                                         // Token
	JWTToken       string         `gorm:"type:text" column:"jwt_token" json:"-"`                                      // JWT Token
	Remark         string         `gorm:"size:200" json:"remark"`                                                     // 备注
	ExchangeTime1  string         `gorm:"column:exchange_time_1;type:time;default:'10:00:00'" json:"exchange_time_1"` // 第一次抢兑时间
	ExchangeTime2  string         `gorm:"column:exchange_time_2;type:time;default:'16:00:00'" json:"exchange_time_2"` // 第二次抢兑时间
	IsActive       bool           `gorm:"default:true" json:"is_active"`                                              // 是否启用
	LastExchangeAt *time.Time     `json:"last_exchange_at,omitempty"`                                                 // 最后抢兑时间
	UpdatedAt      time.Time      `json:"updated_at"`
	CreatedAt      time.Time      `json:"created_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
	User           User           `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Account        Account        `gorm:"foreignKey:AccountID" json:"account,omitempty"`
	Tasks          []ExchangeTask `gorm:"foreignKey:ExchangeAccountID" json:"tasks,omitempty"`
}

func (ExchangeRule) TableName() string {
	return "exchange_rules"
}

// ExchangeAccount 是历史兼容别名；新代码优先使用 ExchangeRule。
type ExchangeAccount = ExchangeRule

// ExchangeTask 抢兑任务模型
type ExchangeTask struct {
	ID                    uint       `gorm:"primarykey" json:"id"`
	SourceOperationID     *string    `gorm:"column:source_operation_id;size:36;uniqueIndex" json:"-"`
	UserID                uint       `gorm:"not null;index" json:"user_id"`                                                     // 用户 ID
	ExchangeAccountID     uint       `gorm:"column:exchange_rule_id;not null;index" json:"exchange_account_id"`                 // 抢兑规则 ID（历史 JSON 字段名保留）
	ProductID             uint       `gorm:"not null;index" json:"product_id"`                                                  // 商品 ID
	PrizeID               string     `gorm:"size:100;not null" json:"prize_id"`                                                 // 商品 ID(来自 API)
	PrizeName             string     `gorm:"size:255;not null" json:"prize_name"`                                               // 商品名称
	TaskType              string     `gorm:"size:20;default:'fixed'" json:"task_type"`                                          // 任务类型：fixed=固定次数，long_term=长期抢兑
	MaxAttempts           int        `gorm:"default:1" json:"max_attempts"`                                                     // 最大抢兑次数 (固定次数模式)
	ScheduledExchangeTime string     `gorm:"column:scheduled_exchange_time;type:time" json:"scheduled_exchange_time,omitempty"` // 任务级指定抢兑时间；为空时使用账号规则时间
	RestockCycle          string     `gorm:"size:20;default:'daily'" json:"restock_cycle"`                                      // 补货周期：daily/weekly/monthly/once
	RestockWeekday        *int       `json:"restock_weekday,omitempty"`                                                         // 周期为 weekly 时的星期，0=周日
	RestockDayOfMonth     *int       `json:"restock_day_of_month,omitempty"`                                                    // 周期为 monthly 时的日期
	RestockTimes          string     `gorm:"type:text" json:"restock_times,omitempty"`                                          // 补货时间点列表，逗号分隔 HH:MM[:SS]
	CustomCron            string     `gorm:"size:120;default:''" json:"custom_cron,omitempty"`                                  // 自定义 cron，优先于固定时间/补货时间
	CalendarPolicy        string     `gorm:"size:20;default:'all'" json:"calendar_policy"`                                      // 日历策略：all/workday/holiday
	HolidayDates          string     `gorm:"type:text" json:"holiday_dates,omitempty"`                                          // 额外节假日 YYYY-MM-DD，逗号分隔
	WorkdayDates          string     `gorm:"type:text" json:"workday_dates,omitempty"`                                          // 调休工作日 YYYY-MM-DD，逗号分隔
	SkipReason            string     `gorm:"size:255;default:''" json:"skip_reason,omitempty"`                                  // 最近一次调度跳过原因
	AttemptedCount        int        `gorm:"default:0" json:"attempted_count"`                                                  // 已抢兑次数
	Status                string     `gorm:"size:20;default:'pending'" json:"status"`                                           // 任务状态
	LastAttemptAt         *time.Time `json:"last_attempt_at,omitempty"`                                                         // 最后抢兑时间
	LastResult            string     `gorm:"type:text" json:"last_result"`                                                      // 最后抢兑结果
	NextRunAt             *time.Time `gorm:"-" json:"next_run_at,omitempty"`                                                    // 下一次预计触发时间（只读预览）
	// 新增字段：优先级、分组、超时控制
	Priority        int              `gorm:"default:0" json:"priority"`            // 优先级：0-普通，1-重要，2-紧急
	TaskGroup       string           `gorm:"size:50;default:''" json:"task_group"` // 任务分组标识
	TimeoutSeconds  int              `gorm:"default:30" json:"timeout_seconds"`    // 超时时间（秒）
	MaxRetries      int              `gorm:"default:3" json:"max_retries"`         // 最大重试次数
	RetryCount      int              `gorm:"default:0" json:"retry_count"`         // 已重试次数
	LastRetryAt     *time.Time       `json:"last_retry_at,omitempty"`              // 最后重试时间
	SuccessCount    int              `gorm:"default:0" json:"success_count"`       // 成功次数
	FailCount       int              `gorm:"default:0" json:"fail_count"`          // 失败次数
	UpdatedAt       time.Time        `json:"updated_at"`
	CreatedAt       time.Time        `json:"created_at"`
	DeletedAt       gorm.DeletedAt   `gorm:"index" json:"-"`
	User            User             `gorm:"foreignKey:UserID" json:"user,omitempty"`
	ExchangeAccount ExchangeAccount  `gorm:"foreignKey:ExchangeAccountID" json:"exchange_account,omitempty"`
	Product         Product          `gorm:"foreignKey:ProductID" json:"product,omitempty"`
	Records         []ExchangeRecord `gorm:"foreignKey:ExchangeTaskID" json:"records,omitempty"`
}

// ExchangeRecord 抢兑记录模型
type ExchangeRecord struct {
	ID                uint      `gorm:"primarykey" json:"id"`
	UserID            uint      `gorm:"not null;index" json:"user_id"`                                     // 用户 ID
	ExchangeAccountID uint      `gorm:"column:exchange_rule_id;not null;index" json:"exchange_account_id"` // 抢兑规则 ID（历史 JSON 字段名保留）
	ExchangeTaskID    *uint     `json:"exchange_task_id,omitempty"`                                        // 抢兑任务 ID
	ProductID         uint      `gorm:"not null;index" json:"product_id"`                                  // 商品 ID
	PrizeID           string    `gorm:"size:100;not null" json:"prize_id"`                                 // 商品 ID(来自 API)
	PrizeName         string    `gorm:"size:255;not null" json:"prize_name"`                               // 商品名称
	Status            string    `gorm:"size:20;not null" json:"status"`                                    // 抢兑结果
	Message           string    `gorm:"type:text" json:"message"`                                          // 抢兑结果消息
	ExecutionTimeMs   int       `gorm:"default:0" json:"execution_time_ms"`                                // 执行时长 (毫秒)
	CreatedAt         time.Time `gorm:"index" json:"created_at"`

	ExchangeAccount *ExchangeAccount `gorm:"foreignKey:ExchangeAccountID" json:"exchange_account,omitempty"`
	Product         *Product         `gorm:"foreignKey:ProductID" json:"product,omitempty"`
}

// CalendarDate 节假日/调休工作日模型。
// date 使用 DATE 主键，day_type 取值 holiday/workday；调度策略优先读取该表，
// 表中未覆盖的日期再回退到周末判断，便于逐年导入真实国务院节假日安排。
type CalendarDate struct {
	Date      string    `gorm:"primaryKey;type:date" json:"date"`
	DayType   string    `gorm:"column:day_type;size:20;not null;index" json:"day_type"`
	Name      string    `gorm:"size:100;default:''" json:"name"`
	Source    string    `gorm:"size:120;default:''" json:"source"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (CalendarDate) TableName() string {
	return "calendar_dates"
}

// ExchangeTaskStatus 抢兑任务状态枚举
type ExchangeTaskStatus string

const (
	ExchangeTaskPending   ExchangeTaskStatus = "pending"
	ExchangeTaskRunning   ExchangeTaskStatus = "running"
	ExchangeTaskCompleted ExchangeTaskStatus = "completed"
	ExchangeTaskFailed    ExchangeTaskStatus = "failed"
)

// ExchangeTaskType 抢兑任务类型枚举
type ExchangeTaskType string

const (
	ExchangeTaskFixed    ExchangeTaskType = "fixed"
	ExchangeTaskLongTerm ExchangeTaskType = "long_term"
)

// ExchangeRecordStatus 抢兑记录状态枚举
type ExchangeRecordStatus string

const (
	ExchangeRecordSuccess ExchangeRecordStatus = "success"
	ExchangeRecordFailed  ExchangeRecordStatus = "failed"
)

// WebSocketMessage WebSocket消息持久化模型
type WebSocketMessage struct {
	ID          uint       `gorm:"primarykey" json:"id"`
	UserID      uint       `gorm:"not null;index;index:idx_ws_user_sequence,priority:1" json:"user_id"`
	MessageID   string     `gorm:"column:message_id;size:64;uniqueIndex" json:"message_id"`
	Sequence    uint64     `gorm:"index:idx_ws_user_sequence,priority:2" json:"sequence"`
	Type        string     `gorm:"size:50;not null" json:"type"`
	Data        string     `gorm:"type:text" json:"data"`
	IsRead      bool       `gorm:"default:false" json:"is_read"`
	IsDelivered bool       `gorm:"default:false" json:"is_delivered"`
	CreatedAt   time.Time  `json:"created_at"`
	ExpiresAt   *time.Time `json:"expires_at,omitempty"`
	ReadAt      *time.Time `json:"read_at,omitempty"`
	DeliveredAt *time.Time `json:"delivered_at,omitempty"`
	AckedAt     *time.Time `json:"acked_at,omitempty"`
}
